python/tskit/jit/numba.py

```diff
@@ -0,0 +1,797 @@
+            rightward tree.
+            The in_range and out_range attributes are updated to reflect the
+            edge changes needed for the current tree when traversing backward.
+            """
+            M = self.ts.num_edges
+            NS = self.ts.num_sites
+            NM = self.ts.num_mutations
+            breakpoints = self.ts.breakpoints
+            right_coords = self.ts.edges_right
+            right_order = self.ts.indexes_edge_removal_order
+            left_coords = self.ts.edges_left
+            left_order = self.ts.indexes_edge_insertion_order
+            sites_position = self.ts.sites_position
+            mutations_site = self.ts.mutations_site
+
+            if self.index == -1:
+                self.index = self.ts.num_trees
+                self.interval = (self.ts.sequence_length, self.interval[1])
+                self.in_range.stop = M - 1
+                self.out_range.stop = M - 1
+                self.direction = REVERSE
+                self.site_range = (NS, NS)
+                self.mutation_range = (NM, NM)
+
+            if self.direction == REVERSE:
+                left_current_index = self.out_range.stop
+                right_current_index = self.in_range.stop
+            else:
+                left_current_index = self.in_range.stop - 1
+                right_current_index = self.out_range.stop - 1
+
+            right = self.interval[0]
+
+            j = left_current_index
+            self.out_range.start = j
+            while j >= 0 and left_coords[left_order[j]] == right:
+                j -= 1
+            self.out_range.stop = j
+            self.out_range.order = left_order
+
+            j = right_current_index
+            self.in_range.start = j
+            while j >= 0 and right_coords[right_order[j]] == right:
+                j -= 1
+            self.in_range.stop = j
+            self.in_range.order = right_order
+
+            self.direction = REVERSE
+            self.index -= 1
+            if self.index == -1:
+                self.set_null()
+            else:
+                left = breakpoints[self.index]
+                self.interval = (left, right)
+
+                # Find sites in current tree interval [left, right) going backward
+                old_site_left, old_site_right = self.site_range
+                j = old_site_left - 1
+                while j >= 0 and sites_position[j] >= left:
+                    j -= 1
+                self.site_range = (j + 1, old_site_left)
+
+                # Find mutations for sites in this interval going backward
+                old_mutation_left, old_mutation_right = self.mutation_range
+                k = old_mutation_left - 1
+                while k >= 0 and mutations_site[k] >= self.site_range[0]:
+                    k -= 1
+                self.mutation_range = (k + 1, old_mutation_left)
+
+            return self.index != -1
+
+    return NumbaTreeSequence
+
+
+def numba_tree_sequence(ts):
```

Comment:
API-wise, I wonder if we're overnaming things here with ``tskit.jit.numba.numba_tree_sequence()`` - it's getting almost Java-esque! How about something like
```python
nts = tskit.numba.jitwrap(ts) # Returns a NumbaTreeSequence
```

"jitwrap" is descriptive and feels like it's something that could in principle be extended to other JIT engines

tree_index = jts.tree_index()

Reply:
Renamed to jitwrap in https://github.com/tskit-dev/tskit/pull/3225/commits/de1d80413d141314c62c244706889107de58acaf